### DISCUSSION TITLE
How does rabbitmq_peer_discovery_k8s plugin work? RabbitMQ Cluster Operator related

### DISCUSSION THREAD

**Discussion by ZichengMa** [rabbitmq/rabbitmq-server#8310] [Questions]
Hi! I am developing a simple rust k8s operator for rabbitmq, and I am confused about how peer discovery works. I found that even if I disabled the `rabbitmq_peer_discovery_k8s plugin`, as long as I set `cluster_formation.peer_discovery_backend = rabbit_peer_discovery_k8s` in the default rabbitmq configmap, the peer discovery works well. If this is true, I don't quite understand the point to have `rabbitmq_peer_discovery_k8s plugin` enabled.
By the way, `rabbitmq_peer_discovery_k8s plugin`  works in a way that it will query k8s API server in order to discover peers. However, a process in the container usually should not be awared of the external environment. Is there other way to discover peers?
Thanks for replying! I am looking forward to your help.

**Comment by michaelklishin** [comment]
You are making claims without explaining how you arrived at the conclusion. For example, how do you observe that "peer discovery still works"? 

If you use a persistent volume for node data, then after the very first cluster formation the nodes [will not perform peer discovery](https://rabbitmq.com/cluster-formation.html#peer-discovery-how-does-it-work). They will contact their known peers. At that point you might as well disable the plugin.

The point of having the plugin enabled is to have the modules in that plugin on the code path of the node. They will be used during the initial cluster formation phase.

RabbitMQ was not the first tool to query the Kubernetes API for peer discovery. Why require the user
to provision Consul or a separate etcd installation (or at the very least a namespace) when
the environment itself has an API perfectly capable of returning a list of peers in a stateful set,
and their state?

The Kubernetes API exists for, well, other applications to use it. RabbitMQ is a stateful distributed system and its needs are very different from a "12 Factor Web app" that outsources all of its state management to external systems. How would a cluster of such system form itself if it was completely
unaware of its peers, or a discovery point of some kind?

There are all [other ways to discover peers](https://rabbitmq.com/cluster-formation.html), and you can implement a tiny interface to add more.
